Add explanation of thread-safety in ROOT (#448)
root-project/web: manual/thread_safetiness/index.md

---

> title: Thread safetiness

Reviewer: ```suggestion
title: Thread safety
```
(google claim safetiness is (for now?) a misspelling)

Author: Oh. But the noun would then be rather `thread safeness`?

Reviewer: I have never seen "safeness" used in this  context (but google says it is used albeit "rarely")

Reviewer 2: May some general remarks from my side.

I would like to include the content in one of the existing files/chapters, for example in https://root.cern/manual/root_classes_data_types_and_global_variables/
Otherwise the Basics part becomes too large. May we must reorganize or rewrite one of the existing chapters, but the aim should be not to let the first part become too overwhelming.

Avoid passive. In the ROOT Manual we talk to the reader directly.
Example: "...objects must not be used concurrently by different threads": Who is the actor? If the user is the actor, write: "Make sure that objects are not used concurrently by different threads".
If the application is the actor, name the actor and write: "xxx ensures that objects are not used concurrently by different threads"".

If a class is mentioned the first time, link to the appropriate page in the Reference Guide. Olivier creates a nice macro {% include ref class="ClassName" %}, for example {% include ref class="TDirectory" %} links to https://root.cern/doc/master/classTDirectory.html

Do not write "TODO" on the page that should be published. We do not indicate a work on progress in the Manual. By default a manual is always work in progress :-)

Reviewer 3: fwiw I agree with @pcanal that the term is [thread safety](https://en.wikipedia.org/wiki/Thread_safety)

Author: @JohannesHentrich Thanks for the comments!  Should be all addressed expect for the place of this section.  I'll leave it to @couet to decide where to put it.

layout: single
sidebar:
  nav: "manual"
toc: true
toc_sticky: true
---

ROOT classes support one of the following thread-safety levels:

> - Thread unsafe: objects must not be used concurrently by different threads, even if objects are not shared among threads.

Reviewer: @axel @jblomer is there a nomenclature to distinguish between the "objects must not be used concurrently by different threads" case (i.e. they have unprotected mutable data member for example) and the case where there is an underlying singleton (static data member of function static variable) that makes that means that it not even safe to use distinct object in different threads?

Author: Isn't this the difference between `conditionally thread safe` and `thread unsafe`?

Or are you looking for a term for classes that don't use unprotected global statics but do contain thread unsafe `const` methods?

Author: @pcanal @Axel-Naumann Speaking of terms, while Wikipedia calls it "conditionally thread safe", I actually did not hear this term before; instead I would have called the concept "thread friendly", which I also find describes the concept better.  What do others think?

Reviewer 4: "Thread friendly" isn't a term of art; "conditionally thread safe" is.

> - Thread unsafe: objects must not be used concurrently by different threads, even if objects are not shared among threads.

Reviewer 3: I think the usage of "object" here is a bit misleading: "objects must not be used concurrently..." implies _the same instance_ while "even if objects are not shared among threads" implies different per-thread instances.

Also, it's worse/more complicated than that: the paragraph mentions generic "objects" and the reader would assume we are talking about instances of one given type, but I in principle you could have race conditions between instances of different types that happen to access the same global state (and since we did not design against that nor we test for that, we have to assume this happens somewhere).

Author: Good point.

> - Conditionally thread safe: objects can be used safely by multiple threads concurrently as long as mutiple threads do not share the same object (i.e. every thread uses a private copy of its objects).  In addition, objects can be shared among threads as long as all threads use only `const` methods of the shared objects.

Reviewer 3: A suggestion that's just to hint at what I think should be made clearer, not meant to be accepted as is:

```suggestion
- Thread-local instances are safe: different threads can safely use different instances of the given type (i.e. every thread uses a local instance of this type).  In addition, objects can be shared among threads as long as all threads use only `const` methods of the shared objects. Instances of different types can freely be used concurrently.
```

- Thread safe: objects can safely be shared and used concurrently by multiple threads.

> Unless otherwise stated, classes starting with `R` (e.g. `RDataFrame`) are conditionally thread safe.

Reviewer 3: ```suggestion
Unless otherwise stated, classes starting with `R` (e.g. `RDataFrame`) offer the basic thread-safety guarantees.
```

> Classes starting with `T` (e.g. `TDirectory`) are thread unsafe.

Reviewer: Actually TDirectory (and as far as I can all the 'core' I/O classes) are "conditionally thread safe".

To find an example of thread unsafe.  An example of thread unsafe class is TUrl since it has const method that modify its content (via mutable data members) without taking any locks.

Reviewer: ```suggestion
Classes starting with `T` (e.g. `TUrl`) should be considered thread unsafe unless otherwise stated.

...

The classes involved in the I/O (e.g. `TDirectory`) are generally conditionally thread safe.
The classes involved in the retrieving of global information (eg. `TROOT`, `TClass`) are generally thread safe.
Most ROOT collections (eg. `TList`, `TObjArray`) are conditionally thread safe and can be turned into thread safe collection by calling `UseRWLock()` (on each given instances).

```

Reviewer 3: "generally" thread safe sounds scary

Reviewer: Well ... english is hard :) the "generally" was meant to bind to the list of classes not the thread safetyness :).  I.e. I was not sure enough to say that "All" the classes involved in the retrieving global information were thread safe.

> Classes starting with `T` (e.g. `TDirectory`) are thread unsafe.

Reviewer 3: maybe:

```suggestion
Types whose name starts with `T` (e.g. `TDirectory`) as well as all globals (such as `gROOT`) are thread unsafe.
```

Reviewer: TDirectory is conditionally thread safe.  gROOT is thread safe. :)

Author: We should add this information to the class documentation of `TDirectory` and `gROOT`.
